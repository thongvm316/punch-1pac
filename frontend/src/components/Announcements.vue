<template>
  <div>
    <div class="toast text-center mt-2"
			:class="[{ 'toast-warning': announcement.status === 'urgent' }, { 'toast-success': announcement.status === 'normal' }]"
			v-for="announcement in headerAnnouncements">
      <button class="btn btn-clear float-right" @click="readAnnouncement(announcement.id)"></button>
      {{ announcement.content }}
    </div>
  </div>
</template>

<script>
import { mapState, mapActions } from 'vuex'

export default {
  name: 'announcements',

  computed: {
    ...mapState('announcements', [
      'headerAnnouncements'
    ])
  },

  methods: {
    ...mapActions('announcements', [
      'readAnnouncement',
      'getHeaderAnnouncements'
    ])
  },

  created () {
    this.getHeaderAnnouncements()
  }
}
</script>
