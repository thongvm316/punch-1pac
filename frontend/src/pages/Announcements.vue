<template>
  <main-layout title="Announcements">
    <table class="table table-hover mt-5 bg-light">
      <thead>
        <tr>
          <th>Title</th>
          <th>Status</th>
          <th>Sent at</th>
        </tr>
      </thead>
      <tbody>
        <router-link
          tag="tr"
          :class="{ 'text-bold-600': !announcement.readed }"
          :to="`/announcements/${announcement.id}`"
          v-for="announcement in announcements"
          :key="announcement.id">
          <td>{{ announcement.title }}</td>
          <td>{{ announcement.status }}</td>
          <td>{{ announcement.created_at | datetime_normal }}</td>
        </router-link>
      </tbody>
    </table>
    <pagination action="getAnnouncements" namespace="announcements" v-if="pager.total_pages > 1"/>
  </main-layout>
</template>

<script>
import MainLayout from '../layouts/Main.vue'
import Pagination from '../components/Pagination.vue'
import { mapState, mapActions } from 'vuex'

export default {
  components: {
    MainLayout,
    Pagination
  },

  computed: {
    ...mapState('announcements', [
      'pager',
      'announcements'
    ])
  },

  methods: {
    ...mapActions('announcements', [
      'getAnnouncements'
    ])
  },

  created () {
    this.getAnnouncements()
  }
}
</script>
