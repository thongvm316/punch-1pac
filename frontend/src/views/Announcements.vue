<template>
  <div>
    <h1>Announcements</h1>
  </div>
</template>
