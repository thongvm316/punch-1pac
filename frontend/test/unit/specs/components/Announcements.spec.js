import { shallowMount } from '@vue/test-utils'
import wrapperOps from '../../supports/wrapper'
import setComputed from '../../supports/set-computed'
import Announcements from '@/components/Announcements'

const readAnnouncement = jest.spyOn(Announcements.methods, 'readAnnouncement')

describe('Announcements.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Announcements, wrapperOps)
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when Announcements was mounted', () => {
    it('should render correctly', () => {
      expect(wrapper.isVueInstance()).toBe(true)
      expect(wrapper.exists()).toBe(true)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when have header announcement', () => {
    it('should dislay normal announcement', () => {
      const announComputed = [
        { status: 'normal', content: 'normal annoucement' }
      ]
      setComputed(wrapper, { headerAnnouncements: announComputed })
      wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
    })

    it('should dislay urgent announcement', () => {
      const announComputed = [
        { status: 'urgent', content: 'urgent annoucement' }
      ]
      setComputed(wrapper, { headerAnnouncements: announComputed })
      wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
    })

    it('should display both announcements', () => {
      const announComputed = [
        { status: 'normal', content: 'normal annoucement' },
        { status: 'urgent', content: 'urgent annoucement' }
      ]
      setComputed(wrapper, { headerAnnouncements: announComputed })
      wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when trigger on read announcement', () => {
    it('should call method', () => {
      const announComputed = [
        { status: 'urgent', content: 'urgent annoucement' }
      ]
      setComputed(wrapper, { headerAnnouncements: announComputed })
      wrapper.find('.toast > button').trigger('click')

      expect(readAnnouncement).toHaveBeenCalled()
    })
  })
})
