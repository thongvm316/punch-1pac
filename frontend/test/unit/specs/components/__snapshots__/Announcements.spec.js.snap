// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Announcements.vue when Announcements was mounted should render correctly 1`] = `<div></div>`;

exports[`Announcements.vue when have header announcement should dislay normal announcement 1`] = `
<div>
  <div class="toast text-center mt-2 toast-success"><button class="btn btn-clear float-right"></button>
    normal annoucement
  </div>
</div>
`;

exports[`Announcements.vue when have header announcement should dislay urgent announcement 1`] = `
<div>
  <div class="toast text-center mt-2 toast-warning"><button class="btn btn-clear float-right"></button>
    urgent annoucement
  </div>
</div>
`;

exports[`Announcements.vue when have header announcement should display both announcements 1`] = `
<div>
  <div class="toast text-center mt-2 toast-success"><button class="btn btn-clear float-right"></button>
    normal annoucement
  </div>
  <div class="toast text-center mt-2 toast-warning"><button class="btn btn-clear float-right"></button>
    urgent annoucement
  </div>
</div>
`;
